### PR TITLE
Fix volume rendering issues

### LINKED ISSUE #428: Volume rendering ignores scalar argument
**Describe the bug, what's wrong, and what you expect:**

When volume rendering, the `scalars` argument is not working and the active scalar array is plotted no matter what.

**To Reproduce**

Run the following code with this file:  [grid.vti.zip](https://github.com/pyvista/pyvista/files/3791473/grid.vti.zip)

```py
import pyvista as pv

grid = pv.read("grid.vti")

p = pv.Plotter(shape=(2,2))
p.add_mesh(grid, scalars="Temperature", cmap="coolwarm")

p.subplot(0,1)
p.add_mesh(grid.copy(), scalars="Temperature_var", cmap="viridis")

p.subplot(1,0)
p.add_volume(grid.copy(), scalars="Temperature", cmap="coolwarm")

p.subplot(1,1)
p.add_volume(grid.copy(), scalars="Temperature_var", cmap="viridis")

p.link_views()
p.view_isometric()
p.show()
```

download

Note the bottom right volume render plot does not match the data above it - its plotting the wrong scalar array

### LINKED ISSUE #430: Volume render opacity doesn't make sense
**Transparent Volumes are Too Bright**

This image is of the same volume with a linear opacity curve, varying the colormap orientation (thermal vs thermal_r). What ends up happening is that even though the opacity values are well below 0.5, perhaps even closer to 0.1, the yellows are way too bright and wash out what should be the more opaque part of the volume. 

On the whole, when manually specifying opacity to the volume render engine, you have to get to incredibly low values to get it to display with transparency. An opacity of 0.5 might as well be solid. 

What's expected is that the transfer function for the opacity should mean that the darker colors are still visible in spite of the lighter, more transparent colors. I also expect that opacity should have more impact on the volume than it currently does - it seems like opacity is too bright overall. 

-----

**To Reproduce**

I can't provide example data, but this is the functional code:
```py
p = pv.Plotter(notebook=False,
               window_size=[3072, 2224], 
               multi_samples=16)

p.add_volume(grid, scalars="vortmag", clim=[1e-03, 1e1], cmap="thermal", opacity="linear", mapper="gpu", n_colors=256*2, ambient=0)
p.camera_position = [(9240.182974604228, 2813.0796919414984, 11182.54953530958),
 (-4091.4200075561566, -2186.2488077169264, 851.0583580728803),
 (-0.4980097002215671, -0.32957308048380274, 0.8021021899394379)]

p.show_grid()
p.show()
p.screenshot("example_1.png")


p = pv.Plotter(notebook=False,
               window_size=[3072, 2224], 
               multi_samples=16)

p.add_volume(grid, scalars="vortmag", clim=[1e-03, 1e1], cmap="thermal_r", opacity="linear", mapper="gpu", n_colors=256*2, ambient=0)
p.camera_position = [(9240.182974604228, 2813.0796919414984, 11182.54953530958),
 (-4091.4200075561566, -2186.2488077169264, 851.0583580728803),
 (-0.4980097002215671, -0.32957308048380274, 0.8021021899394379)]

p.show_grid()
p.show()
p.screenshot("example_2.png")
```

**Screenshots**
Selection_125
Selection_124

-----

**System Information:**
 Please run the following code wherever you are experiencing the bug and paste the output below. This report helps us track down bugs and it is critical to addressing your bug:

```py
# Get system info
import pyvista as pv
print(pv.Report())
```
Selection_126
```txt
# Paste system info here

```

### LINKED ISSUE #431: Volume rendering opacity is not working as expected
DUPLICATE OF #430 

**Describe the bug, what's wrong, and what you expect:**

The opacity mapping when volume rendering is not working as expected.

-----

**To Reproduce**

This dataset:  [grid.vti.zip](https://github.com/pyvista/pyvista/files/3791546/grid.vti.zip)

The yellow in the chosen colormap should be nearly transparent but it is very saturated - look at the colorbar and what is shown - they do not match.

```py
import pyvista as pv
from pyvista import examples

grid = pv.read("grid.vti")

p = pv.Plotter(notebook=0)
p.add_volume(grid, scalars="Temperature", 
             cmap="thermal", opacity="linear_r")
p.show()
```

**Screenshots**

### LINKED ISSUE #444: Empty volume rendering caused by spacing choice
## Description
I trying to do volume rendering with some data that I have in RectilinearGrid format. To do that I'm trying first to create a UniformGrid object. 
```
s = pv.read("reciprocal_total.vtk")
U = pv.UniformGrid()
U.origin = [-0.1, -0.1, -0.1]
U.spacing = [0.2/29, 0.2/29, 0.2/29]
U.dimensions = [ 30, 30, 30 ]
U['DT'] = s['DT_tot']
U.plot(volume=True,notebook=False,opacity='linear_r')
```
However when I ask it to plot it I get an empty image:
empty
If I plot it as a surface

```
p = pv.Plotter(notebook=False, shape=(1, 2), border=False)
p.subplot(0, 0)
p.add_text("Rectilinear", font_size=24)
p.add_mesh(s)
p.subplot(0, 1)
p.add_text("Uniform", font_size=24)
p.add_mesh(U)
p.link_views()
p.show()
```
 it works, and I get the same result as if I plotted the rectilinear grid: 
surface_works

If I extract the array into a n-dimensional numpy array 
```
U2 = np.reshape(U.get_array('DT'),[30,30,30],'F')
p = pv.Plotter(notebook=False)
p.add_volume(U2,resolution=[100,100,100], opacity='sigmoid_r')
p.show()
```
the volume rendering works (not very pretty but it works) so I guess I am doing something wrong:
it_works

What is what I'm doing wrong in the first code?

## Example Data
The sample data can be found at [reciprocal_total.vtk](https://drive.google.com/open?id=1fNcfrdVrjuc-VNlo2vVWfEslz4JYzOpm)

### PR DESCRIPTION
**Pull request by banesullivan**
Resolve #428, resolve #430, resolve #431, resolve #444

Woot! 🎉 

This fixes all the opacity issues we've been having when volume rendering. The problem had to do with how the opacity transfer function is applied over a distance - the default distance to apply it over was set to 1.0 which means if you have a mesh with an extent less than 1.0, the whole dataset could be transparent (#444) as the opacity function would be applied over an extent bigger than the mesh. Likewise, if you have a really big mesh, you wouldn't see any effect from the opacity transfer function (#430 and #431) for the inverse reason.

This also:
- fixes an issue around how the scalars were not properly set when volume rendering (#428)
- implements an option for shading when volume rendering
- adds diffuse and specular controls to volume rendering

@keltonhalbert, can you check this out with your data and make sure all is good?

@pyvista/developers, I'd like to merge this as soon as possible so a review would be greatly appreciated! Notable changes

- changes default mapper to `"smart"`. We've gone back and forth between what mapper to use in a few issues and implemented the `"fixed_point"` mapper at first as outlined in https://github.com/pyvista/pyvista/pull/231#issuecomment-496247649 and later justified in https://github.com/pyvista/pyvista/issues/301#issuecomment-509709765, but now I am having significant performance issues when using that mapper on my machine. The smart mapper performs the best for me so that's why I propose switching it. @supersubscript, do you have any objections to this? This may need to be checked for support on Windows and Linux (I'm Mac).
- Adds controls to two attributes of the volume property (minor impact) but are my default values good? for the opacity distance, I chose 1/100th of the length of the mesh which is what ParaView implements (I think).

------

## Fix to #428 

download

## Fix to #430 and #431 

download

## Shading

sphx_glr_volume_004

### CODE REVIEW

**Comment by banesullivan** [comment]
Okay so it looks like the `"smart"` mapper had an issue on the Windows testing pipeline, so I put the mapper choice in `rcParams` with a default value of `"smart"` unless on windows, then `"fixed_point"` is used.

**Comment by banesullivan** [comment]
After this PR, I want to totally refactor `add_mesh` and `add_volume` methods as they take a lot of the same keyword arguments and there is a lot of duplicated code between them. It would be nice to break them up into a number of shared internal methods

This likely won't happen before `0.23.0`

**Comment by banesullivan** [comment]
I thought I'd look into default volume rendering between PyVista and ParaView - overall, we're looking good. I still can't put my finger on exactly how ParaView decides the scalar opacity unit distance for each data set - but I'm searching the ParaView code as their default choice is always really good.

------

However, I encountered some strange behavior on ParaView's side... ParaView appears to resample cell data to the points when volume rendering.

For example, here is the default handling of a sample mesh with cell data in PyVista (left) and those data resampled to the points (right):

**Use this data:  [grid_50.vti.zip](https://github.com/pyvista/pyvista/files/3890092/grid_50.vti.zip)**

```py
import pyvista as pv
pv.set_plot_theme("paraview")

grid_c = pv.read("grid_50.vti")
grid_p = grid_c.cell_data_to_point_data()

def add_volume(p, vol):
    p.add_volume(vol, cmap="viridis", mapper="smart", 
                 specular_power=100.0, ambient=0, diffuse=0, specular=0,
                 opacity="linear")
    return

p = pv.Plotter(notebook=1, shape=(1,2))
    
add_volume(p, grid_c)
p.subplot(0,1)
add_volume(p, grid_p)

p.link_views()
p.show(cpos=[-1,-1,1])
```

download


-----

Then if we load that data and plot it with default settings in ParaView, it (image below) looks exactly like the version in PyVista where we resampled the data to the points - this makes me wonder if ParaView actually resamples to the points? [Looking in the source code for ParaView](https://github.com/Kitware/ParaView/blob/02b58b44cf567fa703a65aad0dcfd9c8e55463cd/ParaViewCore/ClientServerCore/Rendering/vtkImageVolumeRepresentation.cxx#L301), ParaView should be using the cell data just like we do in PyVista.

Screen Shot 2019-11-25 at 9 47 52 PM

File: pyvista/plotting/plotting.py [hunk 1]
```diff
@@ -1484,14 +1512,19 @@ def add_volume(self, volume, scalars=None, clim=None, resolution=None,
...
+                                    specular_power=specular_power)
 
                 actors.append(a)
             return actors
 
         if not isinstance(volume, pyvista.UniformGrid):
             raise TypeError('Type ({}) not supported for volume rendering at this time. Use `pyvista.UniformGrid`.')
 
+        if opacity_unit_distance is None:
+            opacity_unit_distance = volume.length / (np.mean(volume.dimensions) - 1)
```
Comment (banesullivan) [inline pyvista/plotting/plotting.py]: This is really close to what ParaView does (found by trial and error). I can't find the code for how ParaView chooses this, but for a lot of meshes (when the spacing is 1), this provides the same result as ParaView - I can't figure out exactly how to incorporate the spacing into this equation to get the result that ParaView has.

**Comment by banesullivan** [comment]
Also, there is an interesting setting in ParaView that appears to have overlap with #450 which we may want to look into implementing down the road:

Screen Shot 2019-11-25 at 10 47 07 PM

**Comment by akaszynski** [comment]
@banesullivan this is great work!

**Comment by banesullivan** [comment]
I'm going to go ahead and merge this - we may need a windows user to double-check that the default to `fixed_point` works for them

**Comment by keltonhalbert** [comment]
@banesullivan Sincerest apologies for being delayed in checking this out - great work! I'll see how it looks with my data. I'm actually in the process of using PyVista to do some work around the tops of thunderstorms - so I'll "test in production", as it were. 

As far as refactoring add_mesh and add_volume, I'd like to volunteer myself for those efforts. I think that's something I could help out with. 

I'm not sure how ParaView is doing the unit opacity distance either, but mathematically speaking it should be straightforward to normalize it based on grid spacing to make the results consistent across grid spacings. Perhaps I can contribute by testing some normalization methods. 

I'm intrigued by the differences in point vs cell volume rendering here as well. That's certainly interesting - does it have anything to do with volume shading being enabled?

**Comment by banesullivan** [comment]
> I'm actually in the process of using PyVista to do some work around the tops of thunderstorms - so I'll "test in production", as it were.

🚀 let me know how it goes!

> As far as refactoring add_mesh and add_volume, I'd like to volunteer myself for those efforts. I think that's something I could help out with.

This would be a major help - there is a ton of duplicated code there so it would be great to simplify those two methods as much as possible.

> That's certainly interesting - does it have anything to do with volume shading being enabled?

My first thought was that ParaView was shading it, so I played around with the shading in ParaView and it doesn't seem to be the case - in my post above, the shading is off for PyVista and ParaView

**Pull request merged**